File: templates/python/src/mcp_server/smoke.py
"""Wire-level smoke test.

Spawns the server over stdio against the fake API, then runs initialize,
tools/list, and one tools/call.

Usage: uv run fake-api &  then  uv run smoke
"""

from __future__ import annotations

import os
import sys

import anyio
from mcp.client.session import ClientSession
from mcp.client.stdio import StdioServerParameters, stdio_client


async def run() -> None:
    env = {
        **os.environ,
        "API_BASE_URL": os.environ.get("API_BASE_URL", "http://127.0.0.1:4010"),
        "MCP_TRANSPORT": "stdio",
        "LOG_LEVEL": "WARNING",
    }
    params = StdioServerParameters(command=sys.executable, args=["-m", "mcp_server"], env=env)
    async with stdio_client(params) as (read, write), ClientSession(read, write) as session:
        init = await session.initialize()
        print(f"connected to {init.server_info.name} {init.server_info.version}")

        tools = await session.list_tools()
        names = [t.name for t in tools.tools]
        print("tools:", ", ".join(names))
        if not names:
            raise SystemExit("no tools registered")

        result = await session.call_tool("list_items", {"limit": 2})
        if result.is_error:
            raise SystemExit(f"list_items returned error: {result.content}")
        print("list_items ok:", result.structured_content)

        missing = await session.call_tool("get_item", {"id": "itm_404"})
        if not missing.is_error:
            raise SystemExit("expected get_item on a missing id to return is_error")
        print("get_item error path ok")

        print("SMOKE PASSED")


def main() -> None:
    anyio.run(run)


if __name__ == "__main__":
    main()

File: templates/typescript/scripts/smoke.ts
/**
 * Wire-level smoke test. Spawns the built server over stdio against the fake
 * API, then runs initialize, tools/list, and one tools/call.
 *
 * Usage: npm run build && npm run fake-api &  then  npm run smoke
 */
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { fileURLToPath } from "node:url";
import path from "node:path";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const apiBaseUrl = process.env.API_BASE_URL ?? "http://127.0.0.1:4010";

const transport = new StdioClientTransport({
  command: process.execPath,
  args: [path.join(root, "dist", "index.js")],
  env: { ...process.env, API_BASE_URL: apiBaseUrl, MCP_TRANSPORT: "stdio", LOG_LEVEL: "warn" },
  stderr: "inherit",
});
const client = new Client({ name: "smoke-test", version: "0.0.0" });

try {
  await client.connect(transport);
  const info = client.getServerVersion();
  console.log(`connected to ${info?.name} ${info?.version}`);

  const { tools } = await client.listTools();
  console.log(`tools: ${tools.map((t) => t.name).join(", ")}`);
  if (tools.length === 0) throw new Error("no tools registered");

  const result = await client.callTool({ name: "list_items", arguments: { limit: 2 } });
  if (result.isError) throw new Error(`list_items returned error: ${JSON.stringify(result.content)}`);
  console.log("list_items ok:", JSON.stringify(result.structuredContent));

  const missing = await client.callTool({ name: "get_item", arguments: { id: "itm_404" } });
  if (!missing.isError) throw new Error("expected get_item on a missing id to return isError");
  console.log("get_item error path ok");

  console.log("SMOKE PASSED");
} finally {
  await client.close();
}
